services/blue/src/api/Handlers/AvailabilityHandler.cs

```diff
@@ -22,16 +23,62 @@ public async Task Handle(RequestAvailableRoomTypes message, IMessageHandlerConte
             await using var connection = await _connectionFactory.CreateAsync();
 
             var query = $@"
-SELECT * FROM {V_RoomTypes} AS vrt
+SELECT Id, InternalName, Capacity, `BedType.Id`, `BedType.InternalName`, `Width`, `Length`
+FROM {V_RoomTypes} AS vrt
 WHERE vrt.Id NOT IN (
-	SELECT DISTINCT RoomTypeId FROM Reservations
+	SELECT DISTINCT RoomTypeId FROM {Reservations}
 	WHERE Start >= @Start AND Start <= @End)";
 
-            var queryResult = await connection.ExecuteQueryAsync<RoomType>(query, new { message.Start, message.End });
+            using var reader = await connection.ExecuteReaderAsync(query, new { message.Start, message.End });
 
-            var availableRoomTypes = queryResult.ToArray();
+            var queryModels = ReadQueryModels(reader);
+            var availableRoomTypes = queryModels.Select(QueryModelMapper.Map).ToList();
 
             await context.Reply(new AvailableRoomTypesResponse { RoomTypes = availableRoomTypes });
         }
+
+        private IReadOnlyList<RoomTypeQueryModel> ReadQueryModels(IDataReader reader)
+        {
+            var stash = new Dictionary<string, RoomTypeQueryModel>();
+
+            while (reader.Read())
+            {
+                var roomTypeQueryModel = ReadRoomTypeIfNew();
+
+                var bedTypeQueryModel = new BedTypeQueryModel
+                {
+                    Id = reader.GetString(4),
+                    InternalName = reader.GetString(5),
+                    Width = reader.GetDouble(6),
+                    Length = reader.GetDouble(7),
+                };
+
+                roomTypeQueryModel.BedTypes.Add(bedTypeQueryModel);
+            }
+
+            var queryModels = stash.Values.ToArray();
+            return queryModels;
+
+            // Local functions.
+
+            RoomTypeQueryModel ReadRoomTypeIfNew()
+            {
+                var id = reader.GetString(0);
+
+                if (!stash.TryGetValue(id, out var roomTypeQueryModel))
+                {
+                    roomTypeQueryModel = new RoomTypeQueryModel
```

Comment:
You shouldn't use outer variable here if you then return it.

Reply:
This local function does the following: 'find or create, then return'.
The returned reference will be either the one extracted from the dictionary, or the one just created.

Reply:
I got that, but I find it really confusing that `roomTypeQueryModel` is either assigned in the try-Method or within the if block, and again when the method returns. I would use a local variable and just use the return value, that way it is more clear what happens.